discovery/discovery.go

```diff
@@ -49,27 +49,23 @@ func (o *orchestratorPool) GetOrchestrators(numOrchestrators int) ([]*net.Orches
 	numAvailableOrchs := len(o.uris)
 	numOrchestrators = int(math.Min(float64(numAvailableOrchs), float64(numOrchestrators)))
 	ctx, cancel := context.WithTimeout(context.Background(), getOrchestratorsTimeoutLoop)
-	orchInfos := []*net.OrchestratorInfo{}
-	orchChan := make(chan struct{}, len(o.uris))
-	numResp := 0
-	numSuccessResp := 0
-	respLock := sync.Mutex{}
 
+	infoCh := make(chan *net.OrchestratorInfo, len(o.uris))
+	errCh := make(chan error, len(o.uris))
+	defaultErr := errors.New("unknown error; maybe the predicate failed")
 	getOrchInfo := func(uri *url.URL) {
 		info, err := serverGetOrchInfo(ctx, o.bcast, uri)
-		respLock.Lock()
-		defer respLock.Unlock()
-		numResp++
 		if err == nil && (o.pred == nil || o.pred(info)) {
-			orchInfos = append(orchInfos, info)
-			numSuccessResp++
+			infoCh <- info
+			return
 		}
 		if err != nil && monitor.Enabled {
 			monitor.LogDiscoveryError(err.Error())
 		}
-		if numSuccessResp >= numOrchestrators || numResp >= len(o.uris) {
-			orchChan <- struct{}{}
+		if err == nil {
+			err = defaultErr
```

Comment:
Seems like we might not need `defaultErr` here because:

1. The predicate function [logs an error](https://github.com/livepeer/go-livepeer/blob/56a54051e13889fdd8bfb73011598f631fa4c9f1/discovery/db_discovery.go#L102) when it fails.

2. We are not logging the error received on `errCh`. Logging the error here would add clutter right now since each error would be logged twice since [GetOrchestratorInfo()](https://github.com/livepeer/go-livepeer/blob/56a54051e13889fdd8bfb73011598f631fa4c9f1/server/rpc.go#L215) logs the error before returning it.

Worth considering updating the predicate function and `GetOrchestratorInfo()` to only return the error and let the caller handle logging if needed - out of scope for this PR, but just mentioning it.

Reply:
Good call - this was actually here because I wasn't sure that we could return `nil` from the error channel. I noticed that we didn't actually have tests for this condition, so added some more in https://github.com/livepeer/go-livepeer/pull/1403/commits/dc7828a8a4befdd2a90a6b66498687f404901dec